File: r2c/hello.cpp
#include <iostream>
#include "hello.h"

using std::make_unique;
using std::unique_ptr;

unique_ptr<HelloMsg> new_hello(int msg_num) {
    return make_unique<HelloMsg>(msg_num);
}

int HelloMsg::add_msg_num() {
    ++msg_num;
    std::cout << "HelloMsg::add_msg_num " << msg_num << std::endl;
    return msg_num;
}

File: c2r/hello.rs
pub struct HelloMsg {
    msg_num: i32,
}

impl HelloMsg {
    pub fn new(msg_num: i32) -> HelloMsg {
        HelloMsg {
            msg_num,
        }
    }

    pub fn add_msg_num(&mut self) -> i32 {
        self.msg_num += 1;
        println!("HelloMsg::add_msg_num: {}", self.msg_num);
        self.msg_num
    }
}
